migrations/versions/0425_unique_service_name_2.py

```diff
@@ -14,24 +14,45 @@
 
 
 def upgrade():
-    op.execute("UPDATE services SET normalised_service_name = email_from WHERE normalised_service_name IS NULL")
-    op.execute("UPDATE services_history SET normalised_service_name = email_from WHERE normalised_service_name IS NULL")
-    op.alter_column("services", "normalised_service_name", existing_type=sa.VARCHAR(), nullable=False)
-    op.alter_column("services_history", "normalised_service_name", existing_type=sa.VARCHAR(), nullable=False)
-    op.alter_column("services", "email_from", existing_type=sa.VARCHAR(), nullable=True)
-    op.alter_column("services_history", "email_from", existing_type=sa.VARCHAR(), nullable=True)
-    op.drop_constraint("services_email_from_key", "services", type_="unique")
+    # update services - this transaction will only need row-level write-exclusive locks
+    with op.get_context().autocommit_block():
+        op.execute("UPDATE services SET normalised_service_name = email_from WHERE normalised_service_name IS NULL")
+        op.execute(
+            "UPDATE services_history SET normalised_service_name = email_from WHERE normalised_service_name IS NULL"
+        )
+
+    # this transaction needs an access-exclusive lock on service table.
+    # separate this from the service_history lock so that we reduce the chance of deadlocks
+    with op.get_context().autocommit_block():
+        op.alter_column("services", "normalised_service_name", existing_type=sa.VARCHAR(), nullable=False)
+        op.alter_column("services", "email_from", existing_type=sa.VARCHAR(), nullable=True)
+        op.drop_constraint("services_email_from_key", "services", type_="unique")
+
+    # this transaction needs an access-exclusive lock on service history table
+    with op.get_context().autocommit_block():
+        op.alter_column("services_history", "normalised_service_name", existing_type=sa.VARCHAR(), nullable=False)
+        op.alter_column("services_history", "email_from", existing_type=sa.VARCHAR(), nullable=True)
```

Comment:
I don't believe these will actually be treated as separate transactions FYI - `autocommit_block` puts the postgres connection into an autocommit mode, which means each individual statement is autocommitted. I don't think there's any difference from just having a single autocommit_block context manager around the whole lot.

```
sam:~/work/gds/notifications-api [fix-deadlock]$ flask db upgrade
2023-09-21T16:49:20 app app INFO no-request-id "Logging configured" [in /Users/sam/work/gds/notifications-utils/notifications_utils/logging.py:42]
INFO  [sqlalchemy.engine.Engine] select pg_catalog.version()
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
INFO  [sqlalchemy.engine.Engine] select current_schema()
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
INFO  [sqlalchemy.engine.Engine] show standard_conforming_strings
INFO  [sqlalchemy.engine.Engine] [raw sql] {}
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [sqlalchemy.engine.Engine] select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
INFO  [sqlalchemy.engine.Engine] [generated in 0.00006s] {'name': 'alembic_version'}
INFO  [sqlalchemy.engine.Engine] SELECT alembic_version.version_num
FROM alembic_version
INFO  [sqlalchemy.engine.Engine] [generated in 0.00007s] {}
INFO  [alembic.runtime.migration] Running upgrade 0424_n_history_created_at -> 0425_unique_service_name_2, Revision ID: 0425_unique_service_name_2
Revises: 0424_n_history_created_at
Create Date: 2023-09-01 10:10:19.011060
INFO  [sqlalchemy.engine.Engine] COMMIT
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO  [sqlalchemy.engine.Engine] UPDATE services SET normalised_service_name = email_from WHERE normalised_service_name IS NULL
INFO  [sqlalchemy.engine.Engine] [generated in 0.00006s] {}
INFO  [sqlalchemy.engine.Engine] UPDATE services_history SET normalised_service_name = email_from WHERE normalised_service_name IS NULL
INFO  [sqlalchemy.engine.Engine] [generated in 0.00005s] {}
INFO  [sqlalchemy.engine.Engine] COMMIT using DBAPI connection.commit(), DBAPI should ignore due to autocommit mode
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [sqlalchemy.engine.Engine] COMMIT
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO  [sqlalchemy.engine.Engine] ALTER TABLE services ALTER COLUMN normalised_service_name SET NOT NULL
INFO  [sqlalchemy.engine.Engine] [no key 0.00004s] {}
INFO  [sqlalchemy.engine.Engine] ALTER TABLE services ALTER COLUMN email_from DROP NOT NULL
INFO  [sqlalchemy.engine.Engine] [no key 0.00004s] {}
INFO  [sqlalchemy.engine.Engine] ALTER TABLE services DROP CONSTRAINT services_email_from_key
INFO  [sqlalchemy.engine.Engine] [no key 0.00004s] {}
INFO  [sqlalchemy.engine.Engine] COMMIT using DBAPI connection.commit(), DBAPI should ignore due to autocommit mode
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [sqlalchemy.engine.Engine] COMMIT
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit; DBAPI should not BEGIN due to autocommit mode)
INFO  [sqlalchemy.engine.Engine] ALTER TABLE services_history ALTER COLUMN normalised_service_name SET NOT NULL
INFO  [sqlalchemy.engine.Engine] [no key 0.00003s] {}
INFO  [sqlalchemy.engine.Engine] ALTER TABLE services_history ALTER COLUMN email_from DROP NOT NULL
INFO  [sqlalchemy.engine.Engine] [no key 0.00003s] {}
INFO  [sqlalchemy.engine.Engine] COMMIT using DBAPI connection.commit(), DBAPI should ignore due to autocommit mode
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [sqlalchemy.engine.Engine] UPDATE alembic_version SET version_num='0425_unique_service_name_2' WHERE alembic_version.version_num = '0424_n_history_created_at'
INFO  [sqlalchemy.engine.Engine] [generated in 0.00005s] {}
INFO  [sqlalchemy.engine.Engine] COMMIT
```

see also https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.migration.MigrationContext.autocommit_block

my head currently doesn't want to tell me if this is going to be a problem or not .. but FYI